gobblin-service/src/main/java/org/apache/gobblin/service/modules/utils/ExecutionChecksUtil.java

```diff
@@ -0,0 +1,185 @@
+import org.apache.gobblin.service.modules.flow.SpecCompiler;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.orchestration.TimingEventUtils;
+import org.apache.gobblin.service.modules.orchestration.UserQuotaManager;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+import org.apache.gobblin.service.monitoring.FlowStatusGenerator;
+import org.apache.gobblin.util.ConfigUtils;
+import org.slf4j.Logger;
+
+
+/**
+ * Stateless class with functionality meant to be re-used between the DagManager and Orchestrator when launching
+ * executions of a flow spec. In the common case, the Orchestrator receives a flow to orchestrate, performs necessary
+ * validations, and forwards the execution responsibility to the DagManager. The DagManager's responsibility is to
+ * carry out any flow action requests. However, with launch executions now being stored in the DagActionStateStore, on
+ * restart or leadership change the DagManager has to perform validations before executing any launch actions the
+ * previous leader was unable to complete. Rather than duplicating the code or introducing a circular dependency between
+ * the DagManager and Orchestrator, this class is utilized to store the common functionality. It is stateless and
+ * requires all stateful pieces to be passed as input from the caller.
+ * Note: We expect further refactoring to be done to the DagManager in later stage of multi-active development so we do
+ * not attempt major reorganization as abstractions may change.
+ */
+public final class ExecutionChecksUtil {
+
+  /**
+   * For a given a flowSpec, verifies that an execution is allowed (in case there is an ongoing execution) and the
+   * flowspec can be compiled. If the pre-conditions hold, then a JobExecutionPlan is constructed and returned to the
+   * caller.
+   * @return jobExecutionPlan dag if one can be constructed for the given flowSpec
+   */
+  public static Optional<Dag<JobExecutionPlan>> handleChecksBeforeExecution(
+      SharedFlowMetricsContainer sharedFlowMetricsContainer, SpecCompiler specCompiler, UserQuotaManager quotaManager,
+      Optional<EventSubmitter> eventSubmitter, FlowStatusGenerator flowStatusGenerator, Logger log,
+      boolean flowConcurrencyFlag, FlowSpec flowSpec)
+      throws IOException, InterruptedException {
+    Config flowConfig = flowSpec.getConfig();
+    String flowGroup = flowConfig.getString(ConfigurationKeys.FLOW_GROUP_KEY);
+    String flowName = flowConfig.getString(ConfigurationKeys.FLOW_NAME_KEY);
+    if (!isExecutionPermittedHandler(sharedFlowMetricsContainer, specCompiler, quotaManager, eventSubmitter,
+        flowStatusGenerator, log, flowConcurrencyFlag, flowConfig, flowSpec, flowName, flowGroup)) {
+      return Optional.absent();
+    }
+
+    //Wait for the SpecCompiler to become healthy.
+    specCompiler.awaitHealthy();
+
+    Dag<JobExecutionPlan> jobExecutionPlanDag = specCompiler.compileFlow(flowSpec);
```

Comment:
is this different from what just happened inside `isExecutionPermittedHandler`?  if not, rather than repeat, let's return `Optional<Dag<JobExecutionPlan>>` from the latter, rather than the boolean.  if there's a reason NOT to do that, then please add a comment here, describing the pitfall, so a maintainer gets a clear idea of what to watch out, before potentially trying that